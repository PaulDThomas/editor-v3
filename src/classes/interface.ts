import { EditorV3AtBlock, IEditorV3AtBlock } from "./EditorV3AtBlock";
import { IEditorV3SelectBlock } from "./EditorV3SelectBlock";
import { EditorV3TextBlock, IEditorV3TextBlock } from "./EditorV3TextBlock";
import { IMarkdownSettings } from "./markdown/MarkdownSettings";

/**
 * Available text alignments
 */
export enum EditorV3Align {
  "left" = "left",
  "center" = "center",
  "decimal" = "decimal",
  "right" = "right",
}

/**
 * Object container for all avaialble styles, name: React.CSSProperties
 */
export interface EditorV3Style extends React.CSSProperties {
  isLocked?: boolean;
  isNotAvailabe?: boolean;
}
export interface EditorV3Styles extends Record<string, EditorV3Style> {
  [styleName: string]: EditorV3Style;
}

/**
 * Line object for EditorV3
 */
export interface EditorV3LineProps {
  textBlocks: EditorV3TextBlock[];
}

export interface EditorV3DropListItem<T extends Record<string, string>> {
  text: string;
  data?: T;
  listRender?: HTMLLIElement;
}

/**
 * Text block class objects EditorV3
 */
export type EditorV3BlockClass = EditorV3TextBlock | EditorV3AtBlock;

/**
 * Content object for EditorV3
 * @param textAlignment Alignment of text withing the editor
 * @param decimalAlignPercent Percent of width to align decimal to, 0-100;
 * @param caret Current position of the caret
 * @param styles Available styles
 * @param showMarkdown Print markdown syntax instead of rendering
 * @param markdownSettings Settings for markdown
 */
export interface EditorV3ContentProps {
  allowMarkdown: boolean;
  allowNewLine: boolean;
  allowWindowView: boolean;
  decimalAlignPercent: number;
  markdownSettings: IMarkdownSettings;
  showMarkdown: boolean;
  styles?: EditorV3Styles;
  textAlignment: EditorV3Align;
  atListFunction?: (typedString: string) => Promise<EditorV3DropListItem<Record<string, string>>[]>;
  maxAtListLength: number;
}
export interface EditorV3ContentPropsInput {
  allowMarkdown?: boolean;
  allowNewLine?: boolean;
  allowWindowView?: boolean;
  decimalAlignPercent?: number;
  markdownSettings?: IMarkdownSettings;
  showMarkdown?: boolean;
  styles?: EditorV3Styles;
  textAlignment?: EditorV3Align;
  atListFunction?: (typedString: string) => Promise<EditorV3DropListItem<Record<string, string>>[]>;
  maxAtListLength?: number;
}

/**
 * Line import object for EditorV3
 */
export interface IEditorV3Line {
  textBlocks: (IEditorV3AtBlock | IEditorV3TextBlock | IEditorV3SelectBlock)[];
  contentProps?: EditorV3ContentPropsInput;
}

/**
 * Complete import object for EditorV3
 */
export interface IEditorV3 {
  lines: IEditorV3Line[];
  contentProps?: EditorV3ContentPropsInput;
}

/**
 * Cursor position object for EditorV3
 */
export interface EditorV3Position {
  startLine: number;
  startChar: number;
  endLine: number;
  endChar: number;
  isCollapsed?: boolean;
  focusAt?: "start" | "end";
}
export interface EditorV3PositionF {
  initialLine: number;
  initialChar: number;
  focusLine: number;
  focusChar: number;
}
/**
 * Word positions in the content
 */
export interface EditorV3WordPosition {
  line: number;
  startChar: number;
  endChar: number;
  isLocked: boolean;
}

/**
 * Render properties passed down to elements
 */
export interface EditorV3RenderProps {
  atListFunction?: (typedString: string) => Promise<EditorV3DropListItem<Record<string, string>>[]>;
  caretPosition?: EditorV3Position | null;
  currentEl?: HTMLDivElement | HTMLSpanElement;
  doNotSplitWordSpans?: boolean;
  editableEl?: HTMLDivElement;
  markdownSettings?: IMarkdownSettings;
  maxAtListLength?: number;
  styles?: EditorV3Styles;
}
