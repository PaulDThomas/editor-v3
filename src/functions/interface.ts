// Interfaces
export interface iColourStyles {
  [styleName: string]: React.CSSProperties;
}

export interface iStyleBlock {
  start: number;
  end: number;
  styleName?: string;
}

export interface iColouredLine {
  text: string;
  styles?: iColourStyles;
  styleBlocks?: iStyleBlock[];
}

export enum EditorV3Align {
  'left' = 'left',
  'center' = 'center',
  'decimal' = 'decimal',
  'right' = 'right',
}

export interface AieStyleMap {
  [styleName: string]: { css: React.CSSProperties; aieExclude: string[] };
}
export interface AieStyleExcludeMap {
  [styleName: string]: string[];
}
