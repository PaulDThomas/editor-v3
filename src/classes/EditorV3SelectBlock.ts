import {
  EditorV3TextBlock,
  IEditorV3TextBlock,
  IEditorV3TextBlockOptionalParams,
} from "./EditorV3TextBlock";
import {
  EditorV3DropListItem,
  EditorV3RenderProps,
  EditorV3Style,
  EditorV3WordPosition,
} from "./interface";
import { renderDropdown } from "./toHtml/renderDropdown";

export interface IEditorV3SelectBlockOptionalParams extends IEditorV3TextBlockOptionalParams {
  selectedOption?: string;
  availableOptions?: EditorV3DropListItem<Record<string, string>>[];
}

export interface IEditorV3SelectBlock
  extends IEditorV3TextBlock,
    IEditorV3SelectBlockOptionalParams {}

/**
 * EditorV3SelectBlock - text block that can be one among a list of options
 */
export class EditorV3SelectBlock extends EditorV3TextBlock implements IEditorV3SelectBlock {
  /**
   * Currently selected option
   */
  public selectedOption: string | undefined;

  /**
   * Available options
   */
  public availableOptions: EditorV3DropListItem<Record<string, string>>[] = [];

  get data(): IEditorV3SelectBlock {
    const ret: IEditorV3SelectBlock = super.data;
    ret.selectedOption = this.selectedOption;
    ret.availableOptions = this.availableOptions;
    return ret;
  }

  get wordPositions(): EditorV3WordPosition[] {
    return [
      {
        line: -1,
        startChar: this.lineStartPosition,
        endChar: this.lineStartPosition + this.text.length,
        isLocked: this.isLocked === true,
      },
    ];
  }

  private _styleOrNot = (
    o: EditorV3DropListItem<Record<string, string>>,
  ): EditorV3DropListItem<Record<string, string>> => {
    const option: EditorV3DropListItem<Record<string, string>> = {
      text: o.text,
      data: { text: o.text },
    };
    if (o.data?.style && option.data) option.data.style = o.data.style;
    else if (option.data) option.data.noStyle = "true";
    return option;
  };

  constructor(
    arg: IEditorV3SelectBlock | HTMLSpanElement | DocumentFragment,
    forcedParams?: IEditorV3SelectBlockOptionalParams,
  ) {
    // Usual text block constructor
    super(arg, forcedParams);
    // Force type
    this.type = "select";
    // Check for at specific information
    if (arg instanceof HTMLSpanElement) {
      this.furtherHtml(arg);
    } else if (arg instanceof DocumentFragment) {
      if (arg.childNodes.length !== 1)
        throw new Error("EditorV3SelectBlock:Constructor: DocumentFragment must have 1 child node");
      else if (!(arg.childNodes[0] instanceof HTMLSpanElement))
        throw new Error(
          "EditorV3SelectBlock:Constructor: DocumentFragment child node must be HTMLSpanElement",
        );
      this.furtherHtml(arg.childNodes[0] as HTMLSpanElement);
    } else {
      if (arg.selectedOption) this.selectedOption = arg.selectedOption;
      if (arg.availableOptions) this.availableOptions = arg.availableOptions.map(this._styleOrNot);
    }
    // Force any parameters
    if (forcedParams) {
      if (forcedParams.selectedOption) this.selectedOption = forcedParams.selectedOption;
      if (forcedParams.availableOptions)
        this.availableOptions = forcedParams.availableOptions.map(this._styleOrNot);
    }
  }

  // Read extra data from HTML
  private furtherHtml(arg: HTMLSpanElement) {
    // Copy dataset to selectFactory
    if (arg.dataset.selectedOption) this.selectedOption = arg.dataset.selectedOption;
    if (arg.dataset.availableOptions) {
      this.availableOptions = JSON.parse(arg.dataset.availableOptions);
    }
  }

  // Show dropdown
  public showDropdown = () => {
    this.isSelected = true;
  };

  // Overload html return
  public toHtml(renderProps: EditorV3RenderProps, style?: EditorV3Style): DocumentFragment {
    // Create holder
    const ret = new DocumentFragment();
    if (renderProps.currentEl) renderProps.currentEl.append(ret);
    const span = document.createElement("span");
    ret.appendChild(span);
    span.classList.add("aiev3-tb", "select-block");
    span.dataset.type = "select";

    // Add Text node
    const textNode = document.createTextNode(
      this.text.replaceAll("\u200c", "").replace(/^ /, "\u00a0").replace(/ $/, "\u00a0"),
    );

    span.appendChild(textNode);
    // Add style
    this.applyStyle(span, style);
    // Add label
    if (this.label) span.title = this.label;
    // Delete any existing dropdown on render
    const editor = renderProps.editableEl?.closest(".aiev3") as HTMLDivElement | null;
    if (editor) {
      const existingDropdowns = editor.querySelectorAll("ul.aiev3-dropdown-list");
      existingDropdowns.forEach((dropdown) => dropdown.remove());
    }

    // Add locked status
    if (this.isLocked || !this.isActive) {
      span.classList.add("is-locked");
      span.dataset.isLocked = "true";
      span.contentEditable = "false";
      // Add other data items from the at item
      span.dataset.availableOptions = JSON.stringify(this.availableOptions);
      if (this.selectedOption) span.dataset.selectedOption = this.selectedOption ?? "";

      // Throttle dropdown render
      let renderTimeout: number | null = null;
      if (!renderTimeout && this.isSelected && this.availableOptions.length > 0) {
        span.classList.add("show-dropdown");
        renderTimeout = window.setTimeout(() => {
          renderDropdown(
            span,
            async () => {
              return (
                this.availableOptions.map((option) => {
                  const listItem = document.createElement("li");
                  const tb = new EditorV3TextBlock({
                    text: option.text,
                    style: option.data?.style,
                  });
                  tb.toHtml(
                    { currentEl: listItem },
                    renderProps.styles?.[option.data?.style ?? ""],
                  );
                  return {
                    text: option.text,
                    data: option.data,
                    listRender: listItem,
                  };
                }) ?? []
              );
            },
            Math.max(1, this.availableOptions.length),
            this.text,
          );
        }, 100);
      }
    } else if (this.isActive) {
      span.classList.add("is-active");
    }

    // Add to current element if specified
    if (renderProps.currentEl) {
      renderProps.currentEl.append(ret);
    }
    // Always return
    return ret;
  }
}
