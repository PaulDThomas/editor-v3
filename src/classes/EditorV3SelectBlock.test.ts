import { fireEvent } from "@testing-library/react";
import { EditorV3DropListItem } from "./interface";
import { EditorV3SelectBlock } from "./EditorV3SelectBlock";
import { act } from "react";

describe("EditorV3SelectBlock", () => {
  test("should return a DocumentFragment with the correct structure", () => {
    const text = "-- Select --";
    const style = "bold";
    const availableOptions: EditorV3DropListItem<Record<string, string>>[] = [
      {
        text: "Hello",
      },
      {
        text: "World",
        data: { style: "shiny" },
      },
    ];
    const block = new EditorV3SelectBlock({ text, style, availableOptions });
    expect(block.wordPositions).toEqual([
      {
        line: -1,
        startChar: 0,
        endChar: 12,
        isLocked: true,
      },
    ]);
    block.setActive(true);
    block.showDropdown();
    expect(block.text).toEqual(text);
    const result = block.toHtml({});

    // Check if the DocumentFragment contains a single span element
    expect(result.childNodes.length).toBe(1);
    expect(result.firstChild).toBeInstanceOf(HTMLSpanElement);

    const span = result.firstChild as HTMLSpanElement;

    // Check if the span element has the correct class names
    expect(span.classList.contains("aiev3-tb")).toBe(true);
    expect(span.classList.contains("select-block")).toBe(true);
    expect(span.classList.contains("is-active")).toBe(true);
    expect(span.classList.contains("is-locked")).toBe(true);
    expect(span.classList.contains(`editorv3style-${style}`)).toBe(true);

    // Check if the span element has the correct content and attributes
    expect(span.textContent).toBe(text);
    expect(span.dataset.isLocked).toBe("true");
    expect(span.dataset.styleName).toBe(style);
  });

  test("should correctly read a span element", async () => {
    const span = document.createElement("span");
    span.textContent = "-- Select --";
    span.classList.add("aiev3-tb", "select-block", "editorv3style-bold");
    span.dataset.styleName = "bold";
    span.dataset.isLocked = "true";
    span.dataset.availableOptions = JSON.stringify([
      {
        text: "Hello",
      },
      {
        text: "World",
        data: { style: "shiny" },
      },
    ]);

    const block = new EditorV3SelectBlock(span);
    expect(block.data).toEqual({
      text: "-- Select --",
      type: "select",
      style: "bold",
      availableOptions: [
        {
          text: "Hello",
          data: { noStyle: "true" },
        },
        {
          text: "World",
          data: { style: "shiny" },
        },
      ],
      isLocked: true,
    });
  });

  test("should correctly read a Document fragment", async () => {
    const frag = new DocumentFragment();
    const span = document.createElement("span");
    frag.appendChild(span);
    span.textContent = "-- Select --";
    span.classList.add("aiev3-tb", "select-block", "editorv3style-bold");
    span.title = "Select a number";
    span.dataset.styleName = "bold";
    span.dataset.isLocked = "true";
    span.dataset.availableOptions = JSON.stringify([
      {
        text: "one",
      },
      {
        text: "two",
        data: { style: "shiny" },
      },
      {
        text: "three",
        data: { style: "dull" },
      },
    ]);

    const block = new EditorV3SelectBlock(frag);
    expect(block.data).toEqual({
      text: "-- Select --",
      label: "Select a number",
      type: "select",
      style: "bold",
      availableOptions: [
        {
          text: "one",
          data: { noStyle: "true" },
        },
        {
          text: "two",
          data: { style: "shiny" },
        },
        { text: "three", data: { style: "dull" } },
      ],
      isLocked: true,
    });
  });

  test("should update forcedParams", async () => {
    const text = "-- Select --";
    const style = "bold";
    const availableOptions: EditorV3DropListItem<Record<string, string>>[] = [
      {
        text: "Hello",
      },
      {
        text: "World",
        data: { style: "shiny" },
      },
    ];
    const block = new EditorV3SelectBlock({ text, style, availableOptions });
    expect(block.data).toEqual({
      text,
      type: "select",
      isLocked: true,
      style,
      availableOptions: [
        {
          text: "Hello",
          data: { noStyle: "true" },
        },
        {
          text: "World",
          data: { style: "shiny" },
        },
      ],
    });
    const updatedBlock = new EditorV3SelectBlock(block.data, {
      style: "shiny",
      availableOptions: [
        {
          text: "Hello",
          data: { noStyle: "true" },
        },
        {
          text: "World",
          data: { style: "shiny" },
        },
        { text: "another" },
      ],
    });
    expect(updatedBlock.data).toEqual({
      text,
      type: "select",
      style: "shiny",
      isLocked: true,
      availableOptions: [
        {
          text: "Hello",
          data: { noStyle: "true" },
        },
        {
          text: "World",
          data: { style: "shiny" },
        },
        { text: "another", data: { noStyle: "true" } },
      ],
    });
  });

  test("should return a DocumentFragment when isSelected with the correct structure", () => {
    const text = "-- Select --";
    const label = "Select";
    const style = undefined;
    const availableOptions: EditorV3DropListItem<Record<string, string>>[] = [
      { text: "Hello" },
      {
        text: "World",
        data: { style: "shiny" },
      },
    ];
    const block = new EditorV3SelectBlock({ text, style, availableOptions, label });
    const result = block.toHtml({});

    // Check if the DocumentFragment contains a single span element
    expect(result.childNodes.length).toBe(1);
    expect(result.firstChild).toBeInstanceOf(HTMLSpanElement);

    const span = result.firstChild as HTMLSpanElement;

    // Check if the span element has the correct class names
    expect(span.classList.contains("aiev3-tb")).toBe(true);
    expect(span.classList.contains("select-block")).toBe(true);
    expect(span.classList.contains("is-active")).toBe(false);
    expect(span.classList.contains("show-dropdown")).toBe(false);
    expect(span.classList.contains("is-locked")).toBe(true);
    expect(span.classList.toString().includes("editorv3style")).toBe(false);

    // Check if the span element has the correct content and attributes
    expect(span.textContent).toBe(text);
    expect(span.title).toBe(label);
    expect(span.dataset.isLocked).toBe("true");
    expect(span.dataset.styleName).not.toBeDefined();
  });
});

describe("should return a DocumentFragment with a dropdown", () => {
  test("Ensure previous dropdowns are removed", async () => {
    const text = "-- Select --";
    const style = "bold";
    const availableOptions: EditorV3DropListItem<Record<string, string>>[] = [
      {
        text: "Hello",
      },
      {
        text: "World",
        data: { style: "shiny" },
      },
    ];
    const block = new EditorV3SelectBlock({ text, style, availableOptions });
    block.showDropdown();
    const editor = document.createElement("div");
    document.body.appendChild(editor);
    editor.className = "aiev3";
    const editable = document.createElement("div");
    editable.className = "aiev3-editable";
    editable.contentEditable = "true";
    editor.appendChild(editable);
    const line = document.createElement("div");
    line.className = "aiev3-line";
    editable.appendChild(line);
    // Add an old dropdown
    const oldDropdown = document.createElement("ul");
    oldDropdown.className = "aiev3-dropdown-list";
    oldDropdown.innerHTML = "<li class='aiev3-drop-item'>Old item</li>";
    document.body.appendChild(oldDropdown);

    await act(async () => {
      block.toHtml({
        editableEl: editable,
        currentEl: line,
        caretPosition: { isCollapsed: false, startLine: 0, startChar: 0, endLine: 0, endChar: 10 },
      });
      // Run pending timer from the function to display the dropdown list
      jest.runAllTimers();
    });
    expect(editor.querySelector(".aiev3-dropdown-list")?.textContent).not.toEqual("Old item");
    expect(document.body.innerHTML).toMatchSnapshot();
    // Check dropdown click
    const items = document.body.querySelectorAll(".aiev3-drop-item");
    expect(items).toBeDefined();
    expect(items.length).toBe(2);
    fireEvent.mouseDown(items![1].childNodes[0] as HTMLElement);
    await act(async () => jest.runAllTimers());
    expect(editor.querySelector(".aiev3-drop-item")).toBeNull();
    expect(editor.innerHTML).toMatchSnapshot();
  });

  test("should remove existing style", async () => {
    const text = "-- Select --";
    const style = "bold";
    const availableOptions: EditorV3DropListItem<Record<string, string>>[] = [
      {
        text: "Hello",
      },
      {
        text: "World",
        data: { style: "shiny" },
      },
    ];
    const block = new EditorV3SelectBlock({ text, style, availableOptions });
    block.showDropdown();
    const editor = document.createElement("div");
    document.body.appendChild(editor);
    editor.className = "aiev3";
    const editable = document.createElement("div");
    editable.className = "aiev3-editable";
    editable.contentEditable = "true";
    editor.appendChild(editable);
    const line = document.createElement("div");
    line.className = "aiev3-line";
    editable.appendChild(line);

    await act(async () => {
      block.toHtml({
        editableEl: editable,
        currentEl: line,
        caretPosition: { isCollapsed: false, startLine: 0, startChar: 0, endLine: 0, endChar: 10 },
      });
      // Run pending timer from the function to display the dropdown list
      jest.runAllTimers();
    });
    expect(document.body.innerHTML).toMatchSnapshot();
    expect(document.body.querySelector(".select-block") as HTMLSpanElement).toHaveClass(
      "editorv3style-bold",
    );

    // Check dropdown click
    const items = document.body.querySelectorAll(".aiev3-drop-item");
    expect(items).toBeDefined();
    expect(items.length).toBe(2);
    fireEvent.mouseDown(items![0].childNodes[0] as HTMLElement);
    await act(async () => jest.runAllTimers());
    expect(document.body.querySelector(".aiev3-drop-item")).toBeNull();
    expect(document.body.querySelector(".select-block") as HTMLSpanElement).not.toHaveClass(
      "editorv3style-bold",
    );
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe("EditorV3SelectBlock errors", () => {
  test("should throw an error from an empty document fragment", async () => {
    expect(() => {
      new EditorV3SelectBlock(document.createDocumentFragment());
    }).toThrow("EditorV3SelectBlock:Constructor: DocumentFragment must have 1 child node");
  });
  test("should throw an error from an unknown node type", async () => {
    const frag = new DocumentFragment();
    frag.appendChild(new Text("Bad text!"));
    expect(() => {
      new EditorV3SelectBlock(frag);
    }).toThrow(
      "EditorV3SelectBlock:Constructor: DocumentFragment child node must be HTMLSpanElement",
    );
  });
});

describe("EditorV3SelectBlock markdown output", () => {
  test("should return the correct markdown output", async () => {
    const text = "-- Select --";
    const style = "bold";
    const availableOptions: EditorV3DropListItem<Record<string, string>>[] = [
      {
        text: "Hello",
      },
      {
        text: "World",
        data: { style: "shiny" },
      },
    ];
    const testBlock = new EditorV3SelectBlock({ text, style, availableOptions });
    const result = testBlock.toMarkdown();
    expect(testBlock.toMarkdown()).toEqual("(¬(bold::-- Select --**Hello||shiny::World)¬)");
    const eatOwnTail = new EditorV3SelectBlock(result);
    expect(eatOwnTail.data).toEqual(testBlock.data);
  });
});
