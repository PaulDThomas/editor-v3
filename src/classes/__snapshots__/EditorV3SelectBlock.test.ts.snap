// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return a DocumentFragment with a dropdown Ensure previous dropdowns are removed 1`] = `"<div class="aiev3-editable"><div class="aiev3-line"><span class="aiev3-tb select-block editorv3style-bold is-locked show-dropdown" data-type="select" data-style-name="bold" data-is-locked="true" data-available-options="[{&quot;text&quot;:&quot;Hello&quot;,&quot;data&quot;:{&quot;text&quot;:&quot;Hello&quot;,&quot;noStyle&quot;:&quot;true&quot;}},{&quot;text&quot;:&quot;World&quot;,&quot;data&quot;:{&quot;text&quot;:&quot;World&quot;,&quot;style&quot;:&quot;shiny&quot;}}]">-- Select --</span><ul style="z-index: 1000; left: 0px; top: 0px;" class="aiev3-dropdown-list skip-read"><li class="aiev3-drop-item" data-text="Hello" data-no-style="true"><span class="aiev3-tb">Hello</span></li><li class="aiev3-drop-item" data-text="World" data-style="shiny"><span class="aiev3-tb editorv3style-shiny" data-style-name="shiny">World</span></li></ul></div></div>"`;

exports[`should return a DocumentFragment with a dropdown Ensure previous dropdowns are removed 2`] = `"<div class="aiev3-editable"><div class="aiev3-line"><span class="aiev3-tb select-block editorv3style-bold is-locked show-dropdown editorv3style-shiny" data-type="select" data-style-name="shiny" data-is-locked="true" data-available-options="[{&quot;text&quot;:&quot;Hello&quot;,&quot;data&quot;:{&quot;text&quot;:&quot;Hello&quot;,&quot;noStyle&quot;:&quot;true&quot;}},{&quot;text&quot;:&quot;World&quot;,&quot;data&quot;:{&quot;text&quot;:&quot;World&quot;,&quot;style&quot;:&quot;shiny&quot;}}]" data-style="shiny">World</span></div></div>"`;

exports[`should return a DocumentFragment with a dropdown should remove existing style 1`] = `"<div class="aiev3-editable"><div class="aiev3-line"><span class="aiev3-tb select-block editorv3style-bold is-locked show-dropdown" data-type="select" data-style-name="bold" data-is-locked="true" data-available-options="[{&quot;text&quot;:&quot;Hello&quot;,&quot;data&quot;:{&quot;text&quot;:&quot;Hello&quot;,&quot;noStyle&quot;:&quot;true&quot;}},{&quot;text&quot;:&quot;World&quot;,&quot;data&quot;:{&quot;text&quot;:&quot;World&quot;,&quot;style&quot;:&quot;shiny&quot;}}]">-- Select --</span><ul style="z-index: 1000; left: 0px; top: 0px;" class="aiev3-dropdown-list skip-read"><li class="aiev3-drop-item" data-text="Hello" data-no-style="true"><span class="aiev3-tb">Hello</span></li><li class="aiev3-drop-item" data-text="World" data-style="shiny"><span class="aiev3-tb editorv3style-shiny" data-style-name="shiny">World</span></li></ul></div></div>"`;

exports[`should return a DocumentFragment with a dropdown should remove existing style 2`] = `"<div class="aiev3-editable"><div class="aiev3-line"><span class="aiev3-tb select-block is-locked show-dropdown" data-type="select" data-is-locked="true" data-available-options="[{&quot;text&quot;:&quot;Hello&quot;,&quot;data&quot;:{&quot;text&quot;:&quot;Hello&quot;,&quot;noStyle&quot;:&quot;true&quot;}},{&quot;text&quot;:&quot;World&quot;,&quot;data&quot;:{&quot;text&quot;:&quot;World&quot;,&quot;style&quot;:&quot;shiny&quot;}}]" data-no-style="true">Hello</span></div></div>"`;
